ibis/common/dispatch.py

```diff
@@ -106,3 +108,81 @@ def __call__(self, s: str, *args: Any, **kwargs: Any) -> Any:
     def __doc__(self) -> Any:
         # take the min to give the docstring of the last fallback function
         return min(self.priorities.items(), key=lambda x: x[1])[0].__doc__
+
+
+def lazy_singledispatch(func):
+    """A singledispatch implementation that supports lazily registering
+    implementations."""
+
+    lookup = {object: func}
+    abc_lookup = {}
+    loaders = {}
+
+    def register(cls, func=None):
+        """Registers a new implementation for arguments of type `cls`"""
+
+        def inner(func):
+            if isinstance(cls, tuple):
+                for t in cls:
+                    register(t, func)
+            elif isinstance(cls, abc.ABCMeta):
+                abc_lookup[cls] = func
+            else:
+                lookup[cls] = func
+            return func
+
+        return inner if func is None else inner(func)
+
+    def lazy_loader(module, func=None):
+        """Registers a "loader" function which will be called to register
+        implementations if this dispatch is ever called with arguments types
+        from that module.
+
+        Useful for delaying importing of expensive/optional modules.
+        """
+
+        def inner(func):
+            loaders[module] = func
+            return func
+
+        return inner if func is None else inner(func)
+
+    def dispatch(cls):
+        """Return the implementation for the given `cls`"""
+        for cls2 in cls.__mro__:
+            # 1. Check for a concrete implementation
+            try:
+                impl = lookup[cls2]
+            except KeyError:
+                pass
+            else:
+                if cls is not cls2:
+                    # Cache implementation
+                    lookup[cls] = impl
+                return impl
+            # 2. Check for a lazy loader
+            try:
+                loader = loaders.pop(cls2.__module__.split(".", 1)[0])
+            except KeyError:
+                pass
+            else:
+                loader()
+                return dispatch(cls)
+            # 3. Check for abcs
+            for abc_cls, impl in abc_lookup.items():
+                if issubclass(cls, abc_cls):
+                    lookup[cls] = impl
+                    return impl
+        # Can never get here, since a base `object` implementation is
+        # always registered
+        assert False, "should never get here"  # pragma: no cover
+
+    @functools.wraps(func)
+    def call(arg, *args, **kwargs):
+        return dispatch(type(arg))(arg, *args, **kwargs)
+
+    call.dispatch = dispatch
```

Comment:
Would it make sense to factor this out to a class?

Reply:
This is implemented in the same style as `functools.singledispatch`. A class can work, but brings with it different introspection (and serialization) issues that just using a function and some closures avoids. Would rather keep things this way if possible.

Reply:
Makes sense.